src/ArchimedeanCopula.jl

```diff
@@ -48,45 +48,34 @@ struct ArchimedeanCopula{d,TG} <: Copula{d}
         return new{d,typeof(G)}(G)
     end
 end
-ϕ(C::ArchimedeanCopula{d,TG},t)    where {d,TG} = ϕ(C.G,t)
-ϕ⁻¹(C::ArchimedeanCopula{d,TG},t)  where {d,TG} = ϕ⁻¹(C.G,t)
-ϕ⁽¹⁾(C::ArchimedeanCopula{d,TG},t) where {d,TG} = ϕ⁽¹⁾(C.G,t)
-ϕ⁽ᵏ⁾(C::ArchimedeanCopula{d,TG},t) where {d,TG} = ϕ⁽ᵏ⁾(C.G, Val(d), t)
-williamson_dist(C::ArchimedeanCopula{d,TG}) where {d,TG} = williamson_dist(C.G,Val(d))
+ϕ(C::ArchimedeanCopula{d,TG},t)      where {d,TG} = ϕ(C.G, t)
+ϕ⁻¹(C::ArchimedeanCopula{d,TG},t)    where {d,TG} = ϕ⁻¹(C.G, t)
+ϕ⁻¹⁽¹⁾(C::ArchimedeanCopula{d,TG},t) where {d,TG} = ϕ⁻¹⁽¹⁾(C.G, t)
+ϕ⁽¹⁾(C::ArchimedeanCopula{d,TG},t)   where {d,TG} = ϕ⁽¹⁾(C.G, t)
+ϕ⁽ᵏ⁾(C::ArchimedeanCopula{d,TG},k,t) where {d,TG} = ϕ⁽ᵏ⁾(C.G, k, t)
+williamson_dist(C::ArchimedeanCopula{d,TG}) where {d,TG} = williamson_dist(C.G, Val(d))
 
-
-function _cdf(C::CT,u) where {CT<:ArchimedeanCopula} 
-    sum_ϕ⁻¹u = 0.0
-    for us in u
-        sum_ϕ⁻¹u += ϕ⁻¹(C,us)
-    end
-    return ϕ(C,sum_ϕ⁻¹u)
-end
+_cdf(C::ArchimedeanCopula, u) = ϕ(C, sum(ϕ⁻¹.(C, u)))
 function Distributions._logpdf(C::ArchimedeanCopula{d,TG}, u) where {d,TG}
-    if !all(0 .<= u .<= 1)
+    if !all(0 .< u .< 1)
         return eltype(u)(-Inf)
     end
-    T = promote_type(Float64, eltype(u)) # the Float64 here should be eltype(C) when copulas will be type agnostic... 
-    logdenom = sum_ϕ⁻¹u = zero(T)
-    for us in u
-        ϕ⁻¹u = ϕ⁻¹(C,us)
-        sum_ϕ⁻¹u += ϕ⁻¹u
-        logdenom += log(-ϕ⁽¹⁾(C,ϕ⁻¹u)) # log of negative here because ϕ⁽¹⁾ is necessarily negative
-    end
-    numer = abs(ϕ⁽ᵏ⁾(C, sum_ϕ⁻¹u))
-    if numer > 0
-        r = log(numer) - logdenom
-        !isnan(r) && return r
+    return log(ϕ⁽ᵏ⁾(C, Val(d), sum(ϕ⁻¹.(C, u))) * prod(ϕ⁻¹⁽¹⁾.(C, u)))

```

Comment:
Is'nt that a bit wastefull to compute separately `ϕ⁻¹.(C, u)` and `ϕ⁻¹⁽¹⁾.(C, u)` ? I mean, when there are special methods for both, its certainly OK, but when we go through the default case (automatic derivation) then its wastefull: a call to `WilliamsonTransforms.taylor.(Ref(x -> ϕ⁻¹(C,x)), u , Val(1))` will return the same as `zip(ϕ⁻¹.(C, u), ϕ⁻¹⁽¹⁾.(C, u))`, but with half as much computations: maybe we could try to leverage that ?